go/pkg/expand/expand-group.1.md
# expand-group 1 "" Netspoc "User Manual"

# NAME

expand-group - Substitute group reference by its elements

# SYNOPSIS

expand-group [options] FILE|DIR GROUP-NAME ...

# DESCRIPTION

This program reads a Netspoc configuration and one or more
GROUP-NAMES. It substitutes specified group references in each file
and removes the corresponding group-definition. Each group reference
is substituted by elements of corresponding group definition.
GROUP-NAME is given with type as "group:NAME". Substitution occurs
textual, groups in groups are not expanded. Groups referenced in
intersection or complement are only substituted in simple cases.
If a group can't be expanded at all places, its definition is left unchanged.

Changes are done in place, no backup files are created. But only
changed files are touched.

# OPTIONS

**-f**, **--file** file
:   Read GROUP-NAMES from file.

**-q**, **--quiet**
:   Don't print status messages.

**-h**, **--help**
:   Print a brief help message and exit.

# EXAMPLES

A call to

`expand-group netspoc/ group:g1 group:g2`

would change this Netspoc configuration:

    ## START
    group:g1 = host:a;
    group:g2 = group:g1, host:b;
    group:g3 = group:g2, host:c;
    ## END

like this:

    ## START
    group:g3 =
    host:a,
    host:b,
    host:c,
    ;
    ## END

# COPYRIGHT AND DISCLAIMER

(c) 2025 by Heinz Knutzen, heinz.knutzen@googlemail.com

http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if !, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


Changes
{{$NEXT}}

 - export-netvis: export Dualstack-Address of network
   in field Address2 of node.

6.079     2025-03-17 12:53:54+01:00 Europe/Berlin

 - If service has some owner with name "DA_*",
   this service is marked as visible for all owners with name "DA_*"
   in exported data for Netspoc-Web.
 - Command 'print-path' now accepts dual stack objects.

6.078     2025-02-24 14:42:54+01:00 Europe/Berlin

 - New commandline programs export-netvis and print-path have been added.

    These tools output information as JSON to be used in
    a graph visualization.

    export-netvis creates a list of all networks and routers
    of the given topology with information useful for visualizing
    the topology.

    print-path analyzes the given topology and creates a list of
    names of network nodes (routers and networks) between source and
    destination passed via command line.
    To create the list, print-path uses the same path-walk algorithm
    as used in netspoc main program to create ruleset for devices.

 - Aggregate with IPv4 IP 0.0.0.0/0 no longer matches IPv6 addresses.
   Aggregate with IPv6 IP ::/0 no longer matches IPv4 addresses.
 - Fixed split service from auto intf with IPv4, IPv6
 - Command 'export-netspoc' no longer strips comments and ";"
   from description in details of exported service.

6.077     2025-01-27 11:24:26+01:00 Europe/Berlin

 - Allow dual stack service to have protocols icmp and icmpv6 together.
   Generated rules automatically will use icmp for IPv4 and icmpv6 for IPv6.
 - Removed support for separate IPv6 topology in subdirectory 'ipv6/'.
 - Removed options '--ipv6', '-6'.
 - Removed support for NAT of IPv6 objects.
   After changing attributes of IPv6 addresses from 'ip' to 'ip6',
   supporting NAT for IPv6 would need to change NAT attributes
   to 'ip6', 'hidden6', 'identity6' as well.
   This seems to be superflous, since nobody uses NAT for IPv6.
 - Fixed duplicate hosts in print-group from wrong IPv4 network
   in IPv6 aggregate.
 - Fixed panic in print-group from wrong IPv4 network in IPv6 hosts
   generated by 'auto_ipv6_hosts'.
 - Fixed panic in export-netspoc with rule from IPv4 network to
   combined IPv4/v6 auto interface.
 - Changed 'format-netspoc' to print attributes "ip" and "ip6" in
   one line, if object has no more than 3 attributes.

6.076     2025-01-10 09:39:54+01:00 Europe/Berlin

 - Fixed dual stack bridge. IPv4 and IPv6 interfaces were accidently
   processed together.

6.075     2024-12-19 11:53:37+01:00 Europe/Berlin

 - Fix: Attribute 'policy_distribution_point' at dual stack router
   is only used for matching IPv4/IPv6 part of this router.
 - Fixed command 'export-netspoc'.
   Show IPv4 areas for combined zone. Must not accidently overwrite
   IPv4 areas by IPv6 areas, when combined zones have same name.

6.074     2024-12-16 11:52:23+01:00 Europe/Berlin

 - Fixed wrong error on duplicate interface from router in ipv6/ directory.

6.073     2024-12-16 10:16:33+01:00 Europe/Berlin

 - Added support for dual stack objects, having both, IPv4 and IPv6 addresses.
   This simplifies the modeling of a dual stack toplology.
   It is no longer neccessary to model a separate IPv4 and IPv6 topology.
   Rules between dual stack objects will generate ACls for IPv4 and IPv6.
   Rules between dual stack object and pure IPv4 object will silently ignore
   IPv6 address and generate only ACL for IPv4.
   New attributes have been introduced to define dual stack objects:
   - 'ip6' at network, host, interface and aggregate.
   - 'range6' at host.
   - 'unnumbered6' at network and interface.
   - 'negotiated6' at interface.
   Other changes resulting from use of dual stack objects:
   - If dual stack objects are used as border or inclusive border of an area,
     this defines two areas with identical name:
     one in IPv4 topology and one in IPv6 topology.
   - If dual stack objects are used as interfaces of an pathrestriction,
     this also defines two pathrestrictions in IPv4 and IPv6 topology.
     If the second pathrestriction has only one interface or only interfaces
     outside of a loop, it is silently ignored.
   - New attributes 'ipv4_only' and 'ipv6_only' may be used at service or area.
     This will enable only IPv4 or IPv6 part of dual stack objects.
   - The following attributes are applied only to IPv4 part
     if used in dual stack objects:
     'nat', 'bind_nat', 'subnet_of, 'hub', 'spoke'.
   - Command 'export-netspoc' has been changed to show IPv6 address of
     dual stack objects in attribute 'ip6'.
 - Added new attribute 'auto_ipv6_hosts' to automatically generate
   dual stack hosts from pure IPv4 hosts.
   It will generate IPv6 addresses for hosts
   by combining its IPv4 adress with the IPv6 address of its network.
   These attribute values are provided:
   - auto_ipv6_hosts = readable
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1:172:17:1:48
   - auto_ipv6_hosts = binary
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1::ac11:130
   - auto_ipv6_hosts = none
     No IPv6 address is generated
   This attribute can be used at network, area and host.
 - Fixed command 'cut-netspoc':
   Full path to management_instance is marked now.

6.072     2024-09-05 10:22:13+02:00 Europe/Berlin

 - Fixed command 'remove-from-netspoc'.
   A service is no longer removed accidently, if only some rules of
   a service become empty, but other rules remain.
 - Message "swap objects of 'user' and objects of rules" is shown
   even if no attribute multi_owner is given.
 - Fixed check for NAT network being subnet of some other network.
   If multiple networks have dynamic NAT to same address,
   now a warning is given for each network.
 - Fixed check for useless attribute 'subnet_of'
   where supernet has attribute 'has_subnets'.

6.071     2024-07-29 16:47:58+02:00 Europe/Berlin

 - Fixed generated ACL at router R with managed=local for rule where

go/pkg/expand/expand-group.go
package expand

import (
	"fmt"
	"os"
	"strings"

	"github.com/hknutzen/Netspoc/go/pkg/ast"
	"github.com/hknutzen/Netspoc/go/pkg/astset"
	"github.com/hknutzen/Netspoc/go/pkg/oslink"
	"github.com/spf13/pflag"
)

type state struct {
	*astset.State
	expand   map[string]bool
	elements map[string][]ast.Element
	retain   map[string]bool
	changed  bool
}

func Main(d oslink.Data) int {
	fs := pflag.NewFlagSet(d.Args[0], pflag.ContinueOnError)

	// Setup custom usage function.
	fs.Usage = func() {
		fmt.Fprintf(d.Stderr,
			"Usage: %s [options] FILE|DIR GROUP-NAME ...\n%s",
			d.Args[0], fs.FlagUsages())
	}

	// Command line flags
	quiet := fs.BoolP("quiet", "q", false, "Don't show changed files")
	fromFile := fs.StringP("file", "f", "", "Read GROUP-NAMES from file")
	if err := fs.Parse(d.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			return 1
		}
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		fs.Usage()
		return 1
	}

	// Argument processing
	args := fs.Args()
	if len(args) == 0 {
		fs.Usage()
		return 1
	}
	path := args[0]

	// Names of to be substituted groups.
	var names []string
	if *fromFile != "" {
		l, err := readNames(*fromFile)
		if err != nil {
			fmt.Fprintf(d.Stderr, "Error: %s\n", err)
			return 1
		}
		names = append(names, l...)
	}
	if err := checkNames(args[1:]); err != nil {
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		return 1
	}
	names = append(names, args[1:]...)

	// Change files.
	s := new(state)
	var err error
	s.State, err = astset.Read(path)
	if err != nil {
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		return 1
	}
	err = s.process(names)
	if err != nil {
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		return 1
	}
	s.ShowChanged(d.Stderr, *quiet)
	s.Print()
	return 0
}

func readNames(path string) ([]string, error) {
	var result []string
	bytes, err := os.ReadFile(path)
	if err != nil {
		return result, fmt.Errorf("Can't %s", err)
	}
	result = strings.Fields(string(bytes))
	return result, checkNames(result)
}

func checkNames(names []string) error {
	for _, name := range names {
		if !strings.HasPrefix(name, "group:") {
			return fmt.Errorf("Expected group name but got '%s'", name)
		}
	}
	return nil
}

func (s *state) process(names []string) error {

	s.expand = make(map[string]bool)
	s.elements = make(map[string][]ast.Element)
	s.retain = make(map[string]bool)
	for _, name := range names {
		s.expand[name] = true
	}

	// Collect elements from definitions of to be expanded groups.
	s.Modify(func(n ast.Toplevel) bool {
		name := n.GetName()
		if s.expand[name] {
			group, _ := n.(*ast.TopList)
			s.elements[name] = group.Elements
		}
		return false
	})

	for name := range s.expand {
		if _, found := s.elements[name]; !found {
			return fmt.Errorf("No defintion found for '%s'", name)
		}
	}

	// Repeatedly expand groups in collected elements.
	for name, l := range s.elements {
		for {
			s.changed = false
			s.elementList(&l)
			if !s.changed {
				break
			}
			s.elements[name] = l
		}
	}

	// Expand groups in element lists.
	s.Modify(func(n ast.Toplevel) bool { return s.toplevel(n) })

	// Remove definition of group.
	for name := range s.expand {
		if !s.retain[name] {
			s.DeleteToplevel(name)
		}
	}

	return nil
}

func (s *state) toplevel(n ast.Toplevel) bool {
	s.changed = false
	switch x := n.(type) {
	case *ast.TopList:
		s.elementList(&x.Elements)
	case *ast.Service:
		s.elementList(&x.User.Elements)
		for _, r := range x.Rules {
			s.elementList(&r.Src.Elements)
			s.elementList(&r.Dst.Elements)
		}
	case *ast.Area:
		if x.Border != nil {
			s.elementList(&x.Border.Elements)
		}
		if x.InclusiveBorder != nil {
			s.elementList(&x.InclusiveBorder.Elements)
		}
	}
	return s.changed
}

func (s *state) elementList(l *[]ast.Element) {
	var mod []ast.Element
	for _, n := range *l {
		if vals, ok := s.needExpand(n); ok {
			mod = append(mod, vals...)
			s.changed = true
			continue
		}
		switch obj := n.(type) {
		case *ast.SimpleAuto:
			s.elementList(&obj.Elements)
		case *ast.AggAuto:
			s.elementList(&obj.Elements)
		case *ast.IntfAuto:
			s.elementList(&obj.Elements)
		case *ast.Intersection:
			vals := s.intersection(obj)
			mod = append(mod, vals...)
			continue
		}
		mod = append(mod, n)
	}
	*l = mod
	ast.OrderElements(*l)
}

func (s *state) needExpand(obj ast.Element) ([]ast.Element, bool) {
	if el, ok := obj.(*ast.NamedRef); ok {
		name := el.Type + ":" + el.Name
		if vals, found := s.elements[name]; found {
			return vals, true
		}
	}
	return nil, false
}

func (s *state) intersection(obj *ast.Intersection) []ast.Element {
	// First step:
	// - Split into negated and non negated elements,
	// - substitute negated group by its elements,
	var negated []ast.Element
	var nonNegated []ast.Element
	for _, n := range obj.Elements {
		switch obj := n.(type) {
		case *ast.Complement:
			if vals, ok := s.needExpand(obj.Element); ok {
				s.changed = true
				for _, v := range vals {
					compl := &ast.Complement{Element: v}
					negated = append(negated, compl)
				}
				continue
			}
			negated = append(negated, n)
		default:
			nonNegated = append(nonNegated, n)
		}
	}
	obj.Elements = append(nonNegated, negated...)

	// Second step:
	// If exactly one non negated element was found and if this element
	// is a to be substituted group, then try to expand intersection by
	// intersecting groups elements with negated elements.
	result := []ast.Element{obj}
	if len(nonNegated) != 1 {
		for _, n := range nonNegated {
			s.checkRetain(n)
		}
		return result
	}
	vals, ok := s.needExpand(nonNegated[0])
	if !ok {
		return result
	}
	g := nonNegated[0].(*ast.NamedRef)
	name := g.Type + ":" + g.Name
	retain := true
	if len(vals) == 1 {
		// It is safe to substitute group by its single value,
		// even if intersection can't be expanded below.
		obj.Elements = append(vals, negated...)
		s.changed = true
		retain = false
	}
	for _, n := range negated {
		el1 := n.(*ast.Complement)
		switch compl := el1.Element.(type) {
		default:
			goto NOEXPAND
		case ast.NamedElem:
			t := compl.GetType()
			n := compl.GetName()
			var cp []ast.Element
			for _, v := range vals {
				switch el2 := v.(type) {
				default:
					goto NOEXPAND
				case ast.NamedElem:
					if !(t == el2.GetType() && n == el2.GetName()) {
						cp = append(cp, v)
					}
				}
			}
			if len(cp) == len(vals) {
				// Complement was not found in values of group.
				goto NOEXPAND
			}
			vals = cp
		}
	}
	s.changed = true
	return vals
NOEXPAND:
	s.retain[name] = s.retain[name] || retain
	return result

}

func (s *state) checkRetain(n ast.Element) {
	if obj, ok := n.(*ast.NamedRef); ok {
		name := obj.Type + ":" + obj.Name
		if _, found := s.expand[name]; found {
			s.retain[name] = true
		}
	}
}
